pjnat64/src/pjnat64/pj-nat64.c

```diff
@@ -0,0 +1,1409 @@
+            PJ_LOG(4, (THIS_FILE, "TX patch R-URI addr '%.*s' -> '%.*s' (use mapping)",
+                (int)sip_uri->host.slen, sip_uri->host.ptr,
+                (int)map_addr.slen, map_addr.ptr
+            ));
+            sip_uri->host.slen = map_addr.slen;
+            sip_uri->host.ptr = map_addr.ptr;
+            /* Update patch count */
+            patch_count += 1;
+        }
+    }
+
+    /* Step 2. Patch R-URI with server IPv4 address */
+    if ( 1 ) {
+        /* Step 1. Replace over server IPv4 address */
+        if (hpbx.in4_addr.slen > 0) {
+            PJ_LOG(4, (THIS_FILE, "TX patch R-URI addr use server '%.*s' (use server addr)",
+                (int)hpbx.in4_addr.slen, hpbx.in4_addr.ptr
+            ));
+            sip_uri->host.slen = hpbx.in4_addr.slen;
+            sip_uri->host.ptr = hpbx.in4_addr.ptr;
+            /* Update patch count */
+            patch_count += 1;
+        }
+    }
+
+    /* Step 3. Patch R-URI with custom IPv4 address */
+    if (patch_count == 0) {
+        /* Step 1. Replace over custom IPv4 address */
+        pj_str_t map_addr = { .slen = 8, .ptr = "10.0.0.1" };
+        PJ_LOG(4, (THIS_FILE, "TX patch R-URI addr '%.*s' -> '%.*s' (use 10.0.0.1)",
+            (int)sip_uri->host.slen, sip_uri->host.ptr,
+            (int)map_addr.slen, map_addr.ptr
+        ));
+        sip_uri->host.ptr = map_addr.ptr;
+        sip_uri->host.slen = map_addr.slen;
+        /* Update patch count */
+        patch_count += 1;
+    }
+
+}
+
+static void patch_contact_ipv6_with_ipv4(pjsip_tx_data *tdata, pjsip_msg *msg)
+{
+    pj_pool_t *pool = tdata->pool;
+    unsigned patch_count = 0;
+
+    /* Step 1. Search Contact */
+    pjsip_contact_hdr *contact = (pjsip_contact_hdr*)pjsip_msg_find_hdr(msg, PJSIP_H_CONTACT, NULL);
+    if (contact == NULL) {
+        PJ_LOG(1, (THIS_FILE, "No TX patch Contact"));
+        return;
+    }
+
+#if 0
+    /* Step 1. Common R-URI patch base on IP address mapping */
+    if ((patch_count == 0) && (msg->type == PJSIP_REQUEST_MSG)) {
+        /* Step 1. Get actual R-URI address */
+        pjsip_sip_uri *sip_uri = (pjsip_sip_uri *)pjsip_uri_get_uri(contact->uri);
+        if (sip_uri == NULL) {
+            return;
+        }
+        pj_str_t map_addr = { .slen = 0, .ptr = 0 };
+        /* Step 2. Map IPv6 -> IPv4 address */
+        map_ipv6_with_ipv4(pool, &map_addr, &sip_uri->host);
+        /* Step 3. Mapping routine */
+        if (map_addr.slen > 0) {
+            PJ_LOG(4, (THIS_FILE, "TX patch Contact addr '%.*s':%d -> '%.*s':%d",
+                (int)sip_uri->host.slen, sip_uri->host.ptr, sip_uri->port,
+                (int)map_addr.slen, map_addr.ptr, sip_uri->port
+            ));
+            sip_uri->host.slen = map_addr.slen;
+            sip_uri->host.ptr = map_addr.ptr;
+            /* Update patch count */
+            patch_count += 1;
+        }
+    }
+#endif
+
+    /* Step 2. Patch R-URI on INVITE/ACK/BYE request with server IPv4 address */
```

Comment:
Do you explicitly check INVITE/ACK/BYE? You can do this for every sip request.

Reply:
Yep. I think we can remove checks method here.